# Overall

In [1]:
import pandas as pd
import numpy as np

In [2]:
students_df=pd.read_csv("students_complete.csv")
schools_df=pd.read_csv("schools_complete.csv")
school_data_complete = pd.merge(students_df, schools_df, how="left", on=["school_name", "school_name"])

In [3]:
school_data_complete.head()

Student ID       student_name gender grade        school_name  \
0           0       Paul Bradley      M   9th  Huang High School   
1           1       Victor Smith      M  12th  Huang High School   
2           2    Kevin Rodriguez      M  12th  Huang High School   
3           3  Dr. Richard Scott      M  12th  Huang High School   
4           4         Bonnie Ray      F   9th  Huang High School   

   reading_score  math_score  School ID      type  size   budget  
0             66          79          0  District  2917  1910635  
1             94          61          0  District  2917  1910635  
2             90          60          0  District  2917  1910635  
3             67          58          0  District  2917  1910635  
4             97          84          0  District  2917  1910635

In [4]:
school_data_complete.shape

(39170, 11)

In [5]:
school_data_complete.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 39170 entries, 0 to 39169
Data columns (total 11 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Student ID     39170 non-null  int64 
 1   student_name   39170 non-null  object
 2   gender         39170 non-null  object
 3   grade          39170 non-null  object
 4   school_name    39170 non-null  object
 5   reading_score  39170 non-null  int64 
 6   math_score     39170 non-null  int64 
 7   School ID      39170 non-null  int64 
 8   type           39170 non-null  object
 9   size           39170 non-null  int64 
 10  budget         39170 non-null  int64 
dtypes: int64(6), object(5)
memory usage: 3.6+ MB


In [6]:
school_data_complete.columns

Index(['Student ID', 'student_name', 'gender', 'grade', 'school_name',
       'reading_score', 'math_score', 'School ID', 'type', 'size', 'budget'],
      dtype='object')

In [7]:
list(school_data_complete.columns)

['Student ID',
 'student_name',
 'gender',
 'grade',
 'school_name',
 'reading_score',
 'math_score',
 'School ID',
 'type',
 'size',
 'budget']

In [8]:
school_data_complete.describe()

Student ID  reading_score    math_score     School ID          size  \
count  39170.000000    39170.00000  39170.000000  39170.000000  39170.000000   
mean   19584.500000       81.87784     78.985371      6.978172   3332.957110   
std    11307.549359       10.23958     12.309968      4.444329   1323.914069   
min        0.000000       63.00000     55.000000      0.000000    427.000000   
25%     9792.250000       73.00000     69.000000      3.000000   1858.000000   
50%    19584.500000       82.00000     79.000000      7.000000   2949.000000   
75%    29376.750000       91.00000     89.000000     11.000000   4635.000000   
max    39169.000000       99.00000     99.000000     14.000000   4976.000000   

             budget  
count  3.917000e+04  
mean   2.117241e+06  
std    8.749987e+05  
min    2.480870e+05  
25%    1.081356e+06  
50%    1.910635e+06  
75%    3.022020e+06  
max    3.124928e+06

In [9]:
school_data_complete.head()

Student ID       student_name gender grade        school_name  \
0           0       Paul Bradley      M   9th  Huang High School   
1           1       Victor Smith      M  12th  Huang High School   
2           2    Kevin Rodriguez      M  12th  Huang High School   
3           3  Dr. Richard Scott      M  12th  Huang High School   
4           4         Bonnie Ray      F   9th  Huang High School   

   reading_score  math_score  School ID      type  size   budget  
0             66          79          0  District  2917  1910635  
1             94          61          0  District  2917  1910635  
2             90          60          0  District  2917  1910635  
3             67          58          0  District  2917  1910635  
4             97          84          0  District  2917  1910635

In [10]:
Total_budget = schools_df["budget"].sum()
print(Total_budget)

24649428


In [11]:
Total_school=schools_df["school_name"].count()
print(Total_school)

15


In [12]:
students_df=pd.read_csv("students_complete.csv")
students_df.head()

Student ID       student_name gender grade        school_name  \
0           0       Paul Bradley      M   9th  Huang High School   
1           1       Victor Smith      M  12th  Huang High School   
2           2    Kevin Rodriguez      M  12th  Huang High School   
3           3  Dr. Richard Scott      M  12th  Huang High School   
4           4         Bonnie Ray      F   9th  Huang High School   

   reading_score  math_score  
0             66          79  
1             94          61  
2             90          60  
3             67          58  
4             97          84

In [13]:
Total_student=students_df["student_name"].count()
print(Total_student)

39170


In [14]:
students_df.describe()

Student ID  reading_score    math_score
count  39170.000000    39170.00000  39170.000000
mean   19584.500000       81.87784     78.985371
std    11307.549359       10.23958     12.309968
min        0.000000       63.00000     55.000000
25%     9792.250000       73.00000     69.000000
50%    19584.500000       82.00000     79.000000
75%    29376.750000       91.00000     89.000000
max    39169.000000       99.00000     99.000000

In [15]:
Average_reading_score = round(students_df["reading_score"].mean(),2)
Average_reading_score

81.88

In [16]:
Average_math_score = round(students_df["math_score"].mean(), 2)
Average_math_score

78.99

In [17]:
Passing_reading = round(students_df.loc[students_df["reading_score"]>=70]['reading_score'].count()/students_df.loc[students_df["reading_score"]>=0]['reading_score'].count(), 2)
Passing_reading

0.86

In [18]:
round(len(students_df.loc[students_df.reading_score>=70])/len(students_df.loc[students_df.reading_score>=0]),2)

0.86

In [19]:
Passing_math = round(students_df.loc[students_df["math_score"]>=70]['math_score'].count()/students_df.loc[students_df["math_score"]>=0]['math_score'].count(), 2)
Passing_math

0.75

In [20]:
round(len(students_df.loc[students_df.math_score>=70])/len(students_df.loc[students_df.math_score>=0]),2)

0.75

In [21]:
Percent_overall_pass = round((students_df[(students_df["math_score"]>=70) & (students_df["reading_score"]>=70)]["student_name"].count()/Total_student),2)
Percent_overall_pass

0.65

# Distric Summary

In [22]:
District_Summary= pd.DataFrame({"Total School": [Total_school],
                    "Total Student":[Total_student],
                    "Total Budget": [Total_budget],
                    "Average Math Score": [Average_math_score],
                    "Average Reading Score":[Average_reading_score],
                    "Passing Math":[Passing_math],
                    "Passing Reading": [Passing_reading],
                    "Overall Passing": [Percent_overall_pass]})

District_Sumary= District_Summary[["Total School",
                    "Total Student",
                    "Total Budget",
                    "Average Math Score",
                    "Average Reading Score",
                    "Passing Math",
                    "Passing Reading",
                    "Overall Passing"]]

District_Sumary.style.format({"Total Student": "{:,.0f}",
                    "Total Budget":"${:,.2f}",
                    "Average Math Score": "{:.1f}",
                    "Average Reading Score":"{:.1f}",
                    "Passing Math":"{:.1%}",
                    "Passing Reading":"{:.1%}",
                    "Overall Passing": "{:.1%}"})


# School Summary

In [23]:
BySchool_name = school_data_complete.set_index("school_name").groupby(["school_name"])
School_type = schools_df.set_index("school_name")["type"]

Total_studentper_school = BySchool_name["Student ID"].count()
Total_school_budget = schools_df.set_index("school_name")["budget"]
Per_student_budget = schools_df.set_index("school_name")["budget"]/schools_df.set_index("school_name")["size"]

Average_Math = school_data_complete["math_score"].mean()
Average_Reading = school_data_complete["reading_score"].mean()
Pass_math = school_data_complete[school_data_complete['math_score']>=70].groupby("school_name")["Student ID"].count()/Total_studentper_school
Pass_reading = school_data_complete[school_data_complete['reading_score']>=70].groupby("school_name")["Student ID"].count()/Total_studentper_school
Overall_Pass = school_data_complete[(school_data_complete['math_score']>=70) & (school_data_complete['reading_score']>=70)].groupby("school_name")["Student ID"].count()/Total_studentper_school

School_summary = pd.DataFrame({
                "School type":School_type,
                "Total Students":Total_studentper_school,
                "Total School Budget": Total_school_budget,
                "Per Student Budget": Per_student_budget,
                "Average Math Score":Average_Math,
                "Average Reading Score":Average_Reading,
                "% Passing Math": Pass_math,
                "% Passing Reading":Pass_reading,
                "% Overall Passing":Overall_Pass })

School_summary = School_summary[[
                   "School type",
                    "Total Students",
                    "Total School Budget",
                     "Per Student Budget",
                    "Average Math Score",
                    "Average Reading Score",
                    "% Passing Math",
                    "% Passing Reading",
                     "% Overall Passing"]]

School_summary.head(5).style.format({
                     "Total Students": "{:,.0f}",
                    "Total School Budget": "${:,.0f}",
                     "Per Student Budget": "${:,.0f}",
                    "Average Math Score": "{:,.2f}",
                    "Average Reading Score":"{:,.2f}",
                    "% Passing Math": "{:.1%} ",
                    "% Passing Reading": "{:.1%}",
                     "% Overall Passing":"{:.1%}"})
                         

# Top Performing Schools (By % Overall Passing)

In [24]:
Top_performance = School_summary.sort_values("% Overall Passing", ascending = False)

Top_performance.head().style.format({
                    "Total Students": "{:,}",
                    "Total School Budget": "${:,.0f}",
                     "Per Student Budget": "${:,.0f}",
                    "Average Math Score": "{:,.2f}",
                    "Average Reading Score":"{:,.2f}",
                    "% Passing Math": "{:.1%} ",
                    "% Passing Reading": "{:.1%}",
                     "% Overall Passing":"{:.1%}"})
 

# Bottom Performing Schools (By % Overall Passing)

In [25]:
Bottom_performance = School_summary.sort_values("% Overall Passing", ascending = True)

Bottom_performance.head().style.format({
                    "Total Students": "{:,}",
                    "Total School Budget": "${:,.0f}",
                     "Per Student Budget": "${:,.0f}",
                    "Average Math Score": "{:,.2f}",
                    "Average Reading Score":"{:,.2f}",
                    "% Passing Math": "{:.1%} ",
                    "% Passing Reading": "{:.1%}",
                     "% Overall Passing":"{:.1%}"})

# Math Scores by Grade

In [26]:

nineth_math = students_df.loc[students_df['grade']=="9th"].groupby("school_name")["math_score"].mean()
tenth_math = students_df.loc[students_df['grade']=="10th"].groupby("school_name")["math_score"].mean()
eleventh_math = students_df.loc[students_df['grade']=="11th"].groupby("school_name")["math_score"].mean()
twelth_math = students_df.loc[students_df['grade']=="12th"].groupby("school_name")["math_score"].mean()

math_score_Grade= pd.DataFrame({
                    "9th": nineth_math,
                    "10th": tenth_math,
                    "11th":eleventh_math,
                    "12th": twelth_math})

math_score_Grade.index.name = "School"

math_score_Grade.style.format({
                    "9th": "{:.2f}",
                    "10th":"{:.2f}",
                    "11th":"{:.2f}",
                    "12th": "{:.2f}" })


# Reading Scores by Grade

In [28]:

nineth_reading = students_df.loc[students_df['grade']=="9th"].groupby("school_name")["reading_score"].mean()
tenth_reading = students_df.loc[students_df['grade']=="10th"].groupby("school_name")["reading_score"].mean()
eleventh_reading = students_df.loc[students_df['grade']=="11th"].groupby("school_name")["reading_score"].mean()
twelth_reading = students_df.loc[students_df['grade']=="12th"].groupby("school_name")["reading_score"].mean()

reading_score_Grade= pd.DataFrame({
                    "9th": nineth_reading,
                    "10th": tenth_reading,
                    "11th":eleventh_reading,
                    "12th": twelth_reading})

reading_score_Grade.index.name = "School"

reading_score_Grade.style.format({
                    "9th": "{:.2f}",
                    "10th":"{:.2f}",
                    "11th":"{:.2f}",
                    "12th": "{:.2f}"})


# Scores by School Spending

In [29]:

bins = [0, 584.9999, 629.9999, 644.9999, 680]
group = ["<585", "585-630", "630-645", "645-680"]
school_data_complete["bin_group"] = pd.cut(school_data_complete["budget"]/school_data_complete["size"],bins,labels=group)
group_spending=school_data_complete.groupby("bin_group")

average_math = group_spending["math_score"].mean()
average_read = group_spending["reading_score"].mean()
passing_math = school_data_complete[school_data_complete["math_score"]>=70].groupby("bin_group")["Student ID"].count()/group_spending["Student ID"].count()
passing_reading = school_data_complete[school_data_complete["reading_score"]>=70].groupby("bin_group")["Student ID"].count()/group_spending["Student ID"].count()
Overall_passing=school_data_complete[(school_data_complete["reading_score"]>=70) & (school_data_complete["math_score"]>=70)].groupby("bin_group")["Student ID"].count()/group_spending["Student ID"].count()

Spending_Score_df= pd.DataFrame({
                        "Average Math Score":average_math,
                         "Average Reading Score":average_read,
                        "% Passing Math":passing_math,
                        "% Passing Reading": passing_reading,
                        "% Overall Passing": Overall_passing })

Spending_Score_df.index.name = "Spending ranges per student($)"

Spending_Score_df.style.format({
                        "Average Math Score": "{:.2f}",
                        "Average Reading Score": "{:.2f}",
                        "% Passing Math":"{:.1%}",
                        "% Passing Reading":"{:.1%}",
                        "% Overall Passing": "{:.1%}"})


# Scores by School Size

In [30]:
bins = [0, 999.99, 1999.99, 5000]
group_size = ["Small < 1000", "Medium (1000-2000)", "Large (2000-5000)"]

school_data_complete["school_size"] = pd.cut(school_data_complete["size"], bins,labels=group_size)
school_size_group = school_data_complete.groupby("school_size")

average_math_size = school_size_group["math_score"].mean()
average_read_size = school_size_group["reading_score"].mean()
passing_math_size = school_data_complete[school_data_complete["math_score"]>=70].groupby("school_size")["Student ID"].count()/school_size_group["Student ID"].count()
passing_reading_size = school_data_complete[school_data_complete["reading_score"]>=70].groupby("school_size")["Student ID"].count()/school_size_group["Student ID"].count()
Overall_passing_size = school_data_complete[(school_data_complete["reading_score"]>=70)&(school_data_complete["math_score"]>=70)].groupby("school_size")["Student ID"].count()/school_size_group["Student ID"].count()

Overall_sore_by_size = pd.DataFrame ({
            "Average Math Score": average_math_size,
            "Average Reading Score": average_read_size,
            "% Passing Math": passing_math_size,
            "% Passing Reading": passing_reading_size,
            "% Overall Passing": Overall_passing_size})

Overall_sore_by_size.index.name = "School by size" 

Overall_sore_by_size.style.format({
            "Average Math Score": "{:.2f}",
            "Average Reading Score": "{:.2f}",
            "% Passing Math": "{:.1%}",
            "% Passing Reading": "{:.1%}",
            "% Overall Passing":"{:.1%}"})


# Scores by School Type

In [31]:
school_type = school_data_complete.groupby("type")

average_math_type = school_type["math_score"].mean()
average_reading_type = school_type["reading_score"].mean()
passing_math_type= school_data_complete[school_data_complete["math_score"]>=70].groupby("type")["Student ID"].count()/school_type["type"].count()
passing_reading_type = school_data_complete[school_data_complete["reading_score"]>=70].groupby("type")["Student ID"].count()/school_type["type"].count()
overall_passing_type = school_data_complete[(school_data_complete["math_score"]>=70) & (school_data_complete["reading_score"]>=70)].groupby("type")["Student ID"].count()/school_type["type"].count() 

Over_score_bytype = pd.DataFrame({
                "Average Math Score":average_math_type,
                "Average Reading Score": average_reading_type,
                "% Passing Math":passing_math_type,
                "% Passing Reading":passing_reading_type ,
                "% Overall Passing":overall_passing_type})

Over_score_bytype.index.name= "School type"
        
Over_score_bytype.style.format({
                "Average Math Score": "{:.1f}",
                "Average Reading Score": "{:.1f}",
                "% Passing Math": "{:.1%}",
                "% Passing Reading": "{:.1%}",
                "% Overall Passing": "{:.1%}"})


# Conclusion:

1.	In general, there are 39,170 students in 15 schools, total budget is $24,649,428. Reading average score is 81.9 and math average score is 78.9. 75% of students have reading sore is 91 and math score is 89. 25% of students has math score is 73 and reading score is 69. 75% of students pass Math, and 86% of studens pass Reading, over all 65% pass all.
2.	Top five performing schools are Charter schools, and bottom 5 performing schools are District schools.
3.	 The less spending per student, the more score they get. The less spending group get the highest score (in Math, reading, and over all). The much spending groups get worst scores.
4.	Schools, which have small size, get higher scores. The smallest size school gets highest score. The largest size schools get the worst score.
5.	The charter schools have a better average scores, compared with District schools.
